packages/tldraw/src/lib/shapes/shared/useAsset.ts

```diff
@@ -17,22 +18,68 @@ export function useAsset(shapeId: TLShapeId, assetId: TLAssetId | null, width: n
 		shapeScale,
 	])
 
+	const didAlreadyResolve = useRef(false)
+
+	useEffect(() => {
+		if (url) didAlreadyResolve.current = true
+	}, [url])
+
 	useEffect(() => {
 		if (isCulled) return
 
+		if (assetId && !asset?.props.src) {
+			const preview = editor.getTemporaryAssetPreview(assetId)
+			if (preview) {
+				setUrl(preview.retain())
+				setIsPlaceholder(true)
+				return () => {
+					preview.release()
+				}
+			}
+		}
+
 		let isCancelled = false
-		const timer = editor.timers.setTimeout(async () => {
+
+		async function resolve() {
 			const resolvedUrl = await editor.resolveAssetUrl(assetId, {
 				screenScale,
 			})
-			if (!isCancelled) setUrl(resolvedUrl)
-		}, 500)
+			if (!isCancelled) {
+				setUrl(resolvedUrl)
+				setIsPlaceholder(false)
+			}
+		}
 
-		return () => {
-			clearTimeout(timer)
-			isCancelled = true
+		// if we already resolved the URL, debounce resolution:
+		if (didAlreadyResolve.current) {
+			const timer = editor.timers.setTimeout(resolve, 500)
+
+			return () => {
+				clearTimeout(timer)
+				isCancelled = true
+			}
+		} else {
+			// if not, resolve immediately:
+			resolve()
```

Comment:
there is this issue (i mentioned on Steve's PR), maybe your PR has as well, where the first load of the page will have two network requests unfortunately (this was a discussion we had about adding an `onBeforeMount`, relatedly). we sort of need to at least wait a tick, i think, to avoid this dbl-request...